### PR TITLE
use c++ standard lib that comes with gcc

### PR DESCRIPTION
**Pull request by trns1997**
## Summary
Use C++ standard lib from gcc compiler

### CODE REVIEW

**Comment by trns1997** [comment]
ping @leducp

File: arch/arm/src/common/Toolchain.defs [hunk 1]
```diff
@@ -295,6 +295,10 @@ ARCHCXXFLAGS += -nostdinc++ -Wno-attributes -Wno-unknown-pragmas
 # we can safely disregard warnings of the type "parameter passing for X changed in GCC 7.1."
 # Refer to : https://stackoverflow.com/questions/48149323/what-does-the-gcc-warning-project-parameter-passing-for-x-changed-in-gcc-7-1-m
 
+ifeq ($(CONFIG_GLIBCXX),y)
+  ARCHCXXFLAGS :=$(filter-out -nostdinc++,$(ARCHCXXFLAGS))
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/Toolchain.defs:299]: let's skip the addition of `-nostdinc` in the first place by checking CONFIG_GLIBCXX

**Comment by trns1997** [comment]
Any suggestion on how to overcome this issue:
```
➜  build make
[ 50%] Building CXX object CMakeFiles/basic_example.dir/main.cc.obj
In file included from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/cstdlib:79,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/ext/string_conversions.h:43,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/bits/basic_string.h:4097,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/string:54,
                 from /home/trns1997/nuttxspace/basic_example/main.cc:2:
/home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/stdlib.h:39:3: error: conflicting declaration 'typedef struct div_t div_t'
   39 | } div_t;
      |   ^~~~~
In file included from /home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/nuttx/kmalloc.h:32,
                 from /home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/nuttx/lib/lib.h:31,
                 from /home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/stdio.h:35,
                 from /home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/wchar.h:35,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/cwchar:44,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/bits/postypes.h:40,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/bits/char_traits.h:42,
                 from /home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/c++/13.2.1/string:42:
/home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/stdlib.h:96:22: note: previous declaration as 'typedef struct div_s div_t'
   96 | typedef struct div_s div_t;
      |                      ^~~~~
/home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/stdlib.h:45:3: error: conflicting declaration 'typedef struct ldiv_t ldiv_t'
   45 | } ldiv_t;
      |   ^~~~~~
/home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/stdlib.h:106:23: note: previous declaration as 'typedef struct ldiv_s ldiv_t'
  106 | typedef struct ldiv_s ldiv_t;
      |                       ^~~~~~
/home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/stdlib.h:52:3: error: conflicting declaration 'typedef struct lldiv_t lldiv_t'
   52 | } lldiv_t;
      |   ^~~~~~~
/home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/stdlib.h:116:24: note: previous declaration as 'typedef struct lldiv_s lldiv_t'
  116 | typedef struct lldiv_s lldiv_t;
      |                        ^~~~~~~
/home/trns1997/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi/arm-none-eabi/include/stdlib.h:205:9: error: conflicting declaration of C function 'int putenv(char*)'
  205 | int     putenv (char *__string);
      |         ^~~~~~
/home/trns1997/nuttxspace/basic_example/build/nuttx-export-12.4.0/include/stdlib.h:162:11: note: previous declaration 'int putenv(const char*)'
  162 | int       putenv(FAR const char *string);
      |           ^~~~~~
make[2]: *** [CMakeFiles/basic_example.dir/build.make:76: CMakeFiles/basic_example.dir/main.cc.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/basic_example.dir/all] Error 2
make: *** [Makefile:91: all] Error 2

```
Is there a way to tell the compiler to use nuttx' `stdlib.h` and not the compiler `stdlib.h`. I am guessing we want to use the header file provided by nuttx and not the compiler. For now i just rm nuttx's header file and my project build and runs using the `stdlib.h` from gcc. (Note: I am building an external application using nuttx's export) (Also another note the above error only occurs when trying to include <string> from std)

**Comment by xiaoxiang781216** [comment]
Since nuttx provide libc byself and libc++ use many libc functionality, it's better to integrate gnu libc++ and build with nuttx together like llvm libc++ to avoid this type of error.

File: tools/toolchain.cmake.export [hunk 2]
```diff
@@ -18,14 +18,12 @@ set(CMAKE_C_STANDARD_INCLUDE_DIRECTORIES ${NUTTX_PATH}/include
 set(CMAKE_CXX_STANDARD_INCLUDE_DIRECTORIES
     ${NUTTX_PATH}/include/libcxx ${NUTTX_PATH}/include ${NUTTX_PATH}/arch/chip)
 
-add_compile_options(-nostdlib)
```
Comment (xiaoxiang781216) [inline tools/toolchain.cmake.export:21]: need add conditionally

Comment (trns1997) [reply tools/toolchain.cmake.export:21]: Oh you are right

File: arch/arm/src/common/Toolchain.defs [hunk 3]
```diff
@@ -289,12 +289,16 @@ endif
 # Architecture flags
 
 ARCHCFLAGS += -Wstrict-prototypes -Wno-attributes -Wno-unknown-pragmas
-ARCHCXXFLAGS += -nostdinc++ -Wno-attributes -Wno-unknown-pragmas
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/Toolchain.defs:292]: please change all place which add -nostdinc++

File: include/ctype.h [hunk 4]
```diff
@@ -33,6 +33,15 @@
 #include <nuttx/compiler.h>
 #include <langinfo.h>
 
+#define	_U	01
```
Comment (xiaoxiang781216) [inline include/ctype.h:36]: why need

File: libs/libxx/Kconfig [hunk 5]
```diff
@@ -26,6 +26,11 @@ choice
 	prompt "C++ Library"
 	default LIBCXXMINI
 
+config GLIBCXX
```
Comment (xiaoxiang781216) [inline libs/libxx/Kconfig:29]: LIBCXXTOOLCHAIN

File: libs/libxx/Kconfig [hunk 6]
```diff
@@ -26,6 +26,11 @@ choice
...
+	bool "GCC compiler C++ support"
```
Comment (xiaoxiang781216) [inline libs/libxx/Kconfig:30]:
```suggestion
	bool "Toolchain C++ support"
```

**Comment by xiaoxiang781216** [comment]
@trns1997 please rebase your patch

File: include/ctype.h (hunk 4 above)
Comment (trns1997) [reply include/ctype.h:36]: yup to be removed

File: arch/arm/src/common/Toolchain.defs (hunk 3 above)
Comment (trns1997) [reply arch/arm/src/common/Toolchain.defs:292]: Will do soon

File: arch/arm/src/common/Toolchain.defs [hunk 7]
```diff
@@ -289,12 +289,18 @@ endif
...
+ARCHCXXFLAGS += -Wno-attributes -Wno-unknown-pragmas
 
 # When all C++ code is built using GCC 7.1 or a higher version,
 # we can safely disregard warnings of the type "parameter passing for X changed in GCC 7.1."
 # Refer to : https://stackoverflow.com/questions/48149323/what-does-the-gcc-warning-project-parameter-passing-for-x-changed-in-gcc-7-1-m
 
+ifneq ($(CONFIG_LIBCXXTOOLCHAIN),y)
+  ARCHCXXFLAGS += -nostdinc++
+else
+  ARCHCXXFLAGS += -D_STDLIB_H_
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/Toolchain.defs:301]: why need `-D_STDLIB_H_`?

Comment (trns1997) [reply arch/arm/src/common/Toolchain.defs:301]: This was the simplest work around for the error mentioned above (https://github.com/apache/nuttx/pull/11812#issuecomment-1973470015), the include guard prevents the multiple definition of the stdlib.h header file.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@trns1997 please fix the lint warning:
```
baa1d6092e Use C++ standard lib from toolchain
../nuttx/tools/checkpatch.sh -u -m -g d45a45da94fafeac361c1a4fbad0[24](https://github.com/apache/nuttx/actions/runs/8155014590/job/22289635177?pr=11812#step:3:25)00[25](https://github.com/apache/nuttx/actions/runs/8155014590/job/22289635177?pr=11812#step:3:26)6a0c01..HEAD
ERROR __main__.py:618: Check failed: /home/runner/work/nuttx/nuttx/nuttx/CMakeLists.txt

cmake-format check failed, run following command to update the style:
  $ cmake-format -o <src> <dst>

ERROR __main__.py:618: Check failed: /home/runner/work/nuttx/nuttx/nuttx/CMakeLists.txt
ERROR __main__.py:618: Check failed: /home/runner/work/nuttx/nuttx/nuttx/arch/arm64/src/cmake/Toolchain.cmake
ERROR __main__.py:618: Check failed: /home/runner/work/nuttx/nuttx/nuttx/arch/risc-v/src/cmake/Toolchain.cmake
ERROR __main__.py:618: Check failed: /home/runner/work/nuttx/nuttx/nuttx/arch/x86_64/src/cmake/Toolchain.cmake
```

**Review by leducp — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by davids5** [comment]
@trns1997 What was the reason for this and the benefits and risks of using it?

**Comment by leducp** [comment]
> @trns1997 What was the reason for this and the benefits and risks of using it?

The first reason is an annoying breakage in the llvm libcxx build that appears from time to time (last time: Friday morning on picopi with master). 
The rational is that building libcxx by ourself without their own build system is not really stable: we encountered a lot of trouble with it and swe feel that this is painfull to maintain. For instance, a current problem with the libcxx is a really big binary when no optimization  if a stream is in use (the system link ALL the symbols, even useless one), and missing symbol when optimization is ON.

So we do a peek on ZephyrOS to see how they do it and they just use the compiler C++ library.

Benefits:
- less maintenance
- always up to date library (depending on your compiler version)

Risks:
- no usage so far, so we may encounter regression. I think we need a few months to say if this is really OK or not.

**Pull request merged**